sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -664,6 +669,32 @@ abstract class HadoopFsRelation private[sql](
     })
   }
 
+  /**
+   * Groups the input files by bucket id, if bucketing is enabled and this data source is bucketed.
+   * Returns None if there exists any malformed bucket files.
+   */
+  private def groupBucketFiles(
+      files: Array[FileStatus]): Option[scala.collection.Map[Int, Array[FileStatus]]] = {
+    malformedBucketFile = false
+    if (getBucketSpec.isDefined) {
+      val groupedBucketFiles = mutable.HashMap.empty[Int, mutable.ArrayBuffer[FileStatus]]
+      for (file <- files) {
+        val bucketId = BucketingUtils.getBucketId(file.getPath.getName)
+        if (bucketId.isEmpty) {
+          logError(s"File ${file.getPath} is supposed to be a bucket file, but there is no " +
+            "bucket id information in file name, fallback to non-bucketing mode.")
+          malformedBucketFile = true
+          return None
```

Comment:
you can consider just walking through the data in two passes. the first pass to find the error, and the second pass to do the grouping.